Feature/alternate identifier

marierinden opened the pull request.

marierinden committed:
Add AlternateIdentifierRequest DTO with getter methods

Introduces the AlternateIdentifierRequest class implementing RequestInterface. The class includes properties for name, store, and iri with corresponding getter methods, and a toArray method for serialization.

marierinden committed:
Add AlternateIdentifierResponse DTO implementation

Introduce the `AlternateIdentifierResponse` class under the DTO namespace to represent attribute response data. This class includes properties such as name and store with corresponding getter methods and extends `AbstractResponse` to inherit common response fields.

marierinden committed:
Refactored property name to 'identifierType' to resolve naming conflict with parent class.

marierinden committed:
Add AlternateIdentifierCollectionResponse DTO class

Introduces a new DTO class extending AbstractCollectionResponse for handling collections of AlternateIdentifierResponse objects. This addition supports type consistency and enhances the SDK's response handling capabilities.

marierinden committed:
Add AlternateIdentifierDataMapper implementation

Introduce `AlternateIdentifierDataMapper` to map AlternateIdentifier data between DTOs and arrays. This includes methods for handling single responses, collections, and request transformations, ensuring compatibility with the SDK's structure. Implement error handling for date parsing exceptions.

marierinden committed:
Add AlternateIdentifierResource to support AlternateIdentifier API operations

Introduced `AlternateIdentifierResource` class with methods for CRUD operations on alternateIdentifiers, including validation for required parameters. Updated `ApieraSdk` to include an `alternateIdentifier()` method for accessing the new resource.

fredrik-tveraaen-apiera commented on src/DTO/Request/AlternateIdentifier/AlternateIdentifierCollectionResponse.php:
This must be moved to src/DTO/Response/AlternateIdentifier

fredrik-tveraaen-apiera commented on src/ApieraSdk.php:
```diff
@@ -46,4 +48,13 @@ public function attribute(): AttributeResource
+    /**
+     * @return AlternateIdentifierResource
+     */
+    public function alternateIdentifiers(): AlternateIdentifierResource
```
alternateIdentifiers method must be renamed to alternateIdentifier

fredrik-tveraaen-apiera commented on src/DTO/Request/AlternateIdentifier/AlternateIdentifierCollectionResponse.php:
```diff
@@ -0,0 +1,24 @@
+
+/**
+ * @template-extends AbstractCollectionResponse<AlternateIdentifierResponse>
+ * @author Fredrik Tveraaen <fredrik.tveraaen@apiera.io>
```
Need to correct the @author signature.

fredrik-tveraaen-apiera commented on src/DTO/Request/AlternateIdentifier/AlternateIdentifierRequest.php:
code and type is required and must not be null

fredrik-tveraaen-apiera commented on src/DTO/Request/AlternateIdentifier/AlternateIdentifierRequest.php:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    /**
+     * @return array|mixed[]
```
Change to  @return array<string, mixed>

fredrik-tveraaen-apiera commented on src/DTO/Request/AlternateIdentifier/AlternateIdentifierRequest.php:
```diff
@@ -0,0 +1,60 @@
+ * @package Apiera\Sdk\DTO\Request\Category
+ * @since 0.2.0
+ */
+readonly class AlternateIdentifierRequest implements RequestInterface
```
Add final

fredrik-tveraaen-apiera commented on src/DTO/Response/AlternateIdentifier/AlternateIdentifierResponse.php:
```diff
@@ -0,0 +1,60 @@
+ * @package Apiera\Sdk\DTO\Response\Category
+ * @since 0.2.0
+ */
+readonly class AlternateIdentifierResponse extends AbstractResponse implements ResponseInterface
```
Add final

fredrik-tveraaen-apiera commented on src/DTO/Response/AlternateIdentifier/AlternateIdentifierResponse.php:
```diff
@@ -0,0 +1,60 @@
+     * @param Uuid $uuid
+     * @param DateTimeInterface $createdAt
+     * @param DateTimeInterface $updatedAt
+     * @param string $identifierType
```
Add a comment why we reference type as identifierType

fredrik-tveraaen-apiera commented on src/DTO/Response/AlternateIdentifier/AlternateIdentifierResponse.php:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * @return string
+     */
+    public function getType(): string
```
Rename to getIdentifierType

fredrik-tveraaen-apiera commented on src/DataMapper/AlternateIdentifierDataMapper.php:
```diff
@@ -0,0 +1,83 @@
+        try {
+            return new AlternateIdentifierResponse(
+                id: $responseData['@id'],
+                type: $responseData['@type'],
```
Change to type: LdType::from($responseData['@type']),

fredrik-tveraaen-apiera commented on src/DataMapper/AlternateIdentifierDataMapper.php:
```diff
@@ -0,0 +1,83 @@
+            return new AlternateIdentifierResponse(
+                id: $responseData['@id'],
+                type: $responseData['@type'],
+                uuid: $responseData['uuid'],
```
Change to uuid: Uuid::fromString($responseData['uuid']),

fredrik-tveraaen-apiera commented on src/Resource/AlternateIdentifierResource.php:
```diff
@@ -0,0 +1,134 @@
+ * @package Apiera\Sdk\Resource
+ * @since 0.2.0
+ */
+readonly class AlternateIdentifierResource implements RequestResourceInterface
```
Add final

fredrik-tveraaen-apiera reviewed (changes requested):
Looks good but some changes needed for approval.

marierinden committed:
Refactor based on review feedback

- Moved AlternateIdentifierCollectionResponse.php to Response folder
- Renamed alternateIdentifiers() to alternateIdentifier()
- Updated author signature to correct format
- Made code and type required and non-null
- Changed @return annotation to array<string, mixed>
- Changed readonly classes to final readonly
- Renamed getType() to getIdentifierType() with an explanatory comment
- Updated type and uuid assignments to use LdType::from() and Uuid::fromString()

marierinden committed:
Merge branch 'develop/0.2.0' into feature/alternate-identifier

fredrik-tveraaen-apiera committed:
Add tests for AlternateIdentifierCollectionResponse

Introduce unit tests to verify the functionality and structure of AlternateIdentifierCollectionResponse. Includes tests for constructors, getters, and interface compliance, ensuring proper behavior and type correctness.

fredrik-tveraaen-apiera committed:
Refactor AlternateIdentifierDataMapper and add unit tests

Made the class final, readonly, and updated exception handling to catch broader cases. Simplified code in `fromCollectionResponse` using `array_map` and improved method type safety. Added comprehensive unit tests for better reliability and coverage.

fredrik-tveraaen-apiera committed:
Add unit tests for AlternateIdentifierRequest class

Introduced comprehensive tests covering instantiation, getters, and `toArray` behavior for the `AlternateIdentifierRequest` class. Updated PHPDoc for constructor parameters to improve clarity and maintain consistency.

fredrik-tveraaen-apiera committed:
Add tests for AlternateIdentifierResource with integration support

This commit introduces comprehensive unit and integration tests for the `AlternateIdentifierResource`. Unit tests cover validation, API interactions, and business logic, while integration tests verify end-to-end behavior. Additionally, the endpoint URI in the resource was updated to the correct `/api/v1/alternate_identifiers` path.

fredrik-tveraaen-apiera committed:
Add strict types and unit tests for AlternateIdentifierResponse

Declared strict types in AlternateIdentifierResponse and updated docblock annotations. Added comprehensive unit tests to validate class structure, readonly properties, and behavior of getters and constructor.

fredrik-tveraaen-apiera commented:
Maintainer Note:
As repository maintainers, we're making a one-time exception to bypass the commit signing requirement for this PR. This PR was submitted by our apprentice developer as part of their learning journey with Git and contribution workflows.

While commit signing remains a strict requirement for our repository:
- We're using this as a teaching opportunity
- All future commits will be properly signed
- Our apprentice is now set up with GPG signing for future contributions
